KAFKA-3846: KIP-65: include timestamp in Connect record types

shikhar opened the pull request:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-65%3A+Expose+timestamps+to+Connect

shikhar commented:
@ewencp for review

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/connector/ConnectRecord.java:
```diff
@@ -36,18 +36,19 @@
-        this(topic, kafkaPartition, null, null, valueSchema, value);
-    }
-
-    public ConnectRecord(String topic, Integer kafkaPartition, Schema keySchema, Object key, Schema valueSchema, Object value) {
```
Hmm, I was going to comment that these are public APIs and we need to include the old constructors versions for compatibility. But then I realized that this class is really just an implementation detail which is not currently hidden and in practice nobody should be using these APIs directly.

I'm wondering if maybe we should also consider refactoring some of this into a private `o.a.k.connect.internals` package or something to indicate stuff we really don't want users relying on. It's a bit weird since the `api` artifact was intended to hold the core public APIs, but might still be useful to hide these details.

Not sure this needs to be wrapped into this PR, but maybe some cleanup and could add clarity as to which interfaces we expect to be stable. On the other hand, this class does define some interfaces we expect to be stable in Source and SinkRecord, so maybe hiding it only works for the constructors. Not really sure, just a thought...

hachikuji replied:
ConnectRecord is already abstract and it would be pretty weird for someone to extend it... Also, relocating it to another package would be a compatibility concern, no?

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/sink/SinkRecord.java:
```diff
@@ -18,55 +18,74 @@
     @Override
     public boolean equals(Object o) {
-        if (this == o)
+        if (this == o) {
```
I think your IDE is adding extra braces. (I know this is a bit annoying between Apache Kafka & Confluent projects since they use different styling. iirc default IntelliJ styling works for Kafka if you are using IntelliJ)

shikhar replied:
Ah this is probably because I regenerated equals/hashCode/toString. Will fix.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceRecord.java:
```diff
@@ -46,25 +46,49 @@
+    /**
+     * @deprecated use {@link SourceRecord.Builder}
+     */
+    @Deprecated
```
I think I mentioned this offline, but I'm not sure I would deprecate this yet. I think there are a number of tradeoffs here, this is the set that come to mind immediately:

Pros:
- One builder class, although it has a bunch of per-field methods, will beat the combinatoric explosion of possible constructors when the number of fields increases (as it is now).
- Less error prone for users as they don't have to try to dig out the right constructor to use that has just the right set of fields. If there are any fields with the same type (schema and key/value fields are the ones that come to mind), users may accidentally pass the wrong variable as that argument and not be caught by the compiler.
- Easier compatibility story -- as long as the setter for each field continues to exist in the `Builder`, we're good.

Cons:
- Deprecating relatively new APIs. May be more painful for us to maintain them, but we probably shouldn't be overly aggressive since we want folks to have a stable API to work with.
- Overhead of builder -- I think in a lot of cases this is not a concern as the object is so short lived that collecting it will be very cheap. However, consider that this _can_ be very hot code -- in some cases creating the SourceRecord might be _the majority_ of the work involved in moving messages (e.g. Schema is fixed and a primitive type, so there's not even any other conversion happening). Is the expectation that those use cases would use the single all-fields SourceRecord constructor directly?
- How many variants of the constructor are realistic? We're definitely not expecting to add more fields to messages (that type of change has to come with a _huge_ functionality win to ever make it past KIPs). Also, at least a few fields only make sense when combined (e.g. `valueSchema` + `value`). I'm not sure the number of variants is that high (but maybe I'm wrong -- I'm guessing you raised this as you started copy/pasting all the existing ones to include timestamp).

Heh, not great when my pros/cons list ends up with an equal number of items on either side, but these issues probably warrant further discussion. (I do also very much appreciate keeping an eye out for ways we can cleanup/simplify the API and make the connector developer experience better.)

hachikuji replied:
Just my humble opinion, but I prefer the constructors over the builder since it makes it easy to see which fields are needed and it's generally less verbose. But there are some situations where a Builder may be more helpful, so I think it makes sense to offer that as an alternative.

shikhar replied:
Good points. Especially agree about the 2nd and 3rd con. I'll remove the deprecations and the Builder.

Do you think we should have timestamp variants for all the current constructors, or only a new constructor which takes all args? It feels like specialized usage to me so I think the latter makes more sense.

shikhar replied:
@hachikuji 

> constructors over the builder since it makes it easy to see which fields are needed

I think the way to do that is to have required args on the Builder constructor

(but it seems like best to remove the Builder for now)

hachikuji replied:
@shikhar Sure, but then that also sort of defeats the point of the builder, right?

shikhar replied:
@hachikuji depends... the larger the number of optional parameters, the more the value in Builder as a way to have named & default arguments in Java

ewencp commented on connect/file/src/test/java/org/apache/kafka/connect/file/FileStreamSinkTaskTest.java:
```diff
@@ -19,6 +19,8 @@
 
 import org.apache.kafka.clients.consumer.OffsetAndMetadata;
 import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.record.Record;
```
We'll probably need more in the way of tests for a final version -- validation that the timestamp actually gets passed along for SourceRecords, properly passed into SinkRecords, etc.

shikhar replied:
ack

ewencp commented:
And lets see if this works: @confluentinc/connect might also want to take a look at this.

hachikuji commented on connect/api/src/main/java/org/apache/kafka/connect/sink/SinkRecord.java:
```diff
@@ -18,55 +18,74 @@
 
-    public SinkRecord(String topic, int partition, Schema keySchema, Object key, Schema valueSchema, Object value, long kafkaOffset) {
-        super(topic, partition, keySchema, key, valueSchema, value);
+    public SinkRecord(String topic, int partition,
```
It would probably be helpful to include the old constructor. At least I can imagine users having test cases where they construct SinkRecords directly and it would be nice not to break those unnecessarily.

shikhar replied:
That's fair, then can avoid the `FileStreamSinkTaskTest` change in this diff

shikhar committed:
KAFKA-3846: include timestamp in Connect record types

`SinkRecord` gets `timestampType` and `timestamp`
`SourceRecord` gets `timestamp`

shikhar commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java:
```diff
@@ -348,6 +350,33 @@ public void testWakeupInCommitSyncCausesRetry() throws Exception {
+        workerTask.initialize(TASK_CONFIG);
+        workerTask.initializeAndStart();
+        workerTask.poll(Long.MAX_VALUE);
+        workerTask.poll(Long.MAX_VALUE);
```
Being cargo culty here, I don't actually understand why `poll()` needs to be called twice (otherwise the capture stays empty). @ewencp ?

ewencp replied:
@shikhar This is needed because these tests reuse a common `expectPollInitialAssignment` which _only_ handles getting the member into the group and returns zero records. The subsequent expected poll call will actually retrieve the message and cause the records to be put into the sink task.

shikhar replied:
Ah I see, thanks. I removed the `expectPollInitialAssignment` since that's not under test here, and indeed now only need the single `poll()` call to capture the record.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/sink/SinkRecord.java:
```diff
@@ -18,27 +18,42 @@
 
-    public SinkRecord(String topic, int partition, Schema keySchema, Object key, Schema valueSchema, Object value, long kafkaOffset) {
-        super(topic, partition, keySchema, key, valueSchema, value);
+    public SinkRecord(String topic, int partition, Schema keySchema, Object key, Schema valueSchema, Object value, Long kafkaOffset) {
```
I think `kafkaOffset` was incorrectly changed to `Long`. We'll always have a Kafka offset, so it should be `long`. Also, the current version breaks compatibility since the old signature constructor is no longer available.

shikhar replied:
yikes, good catch

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -359,7 +359,7 @@ private void convertMessages(ConsumerRecords<byte[], byte[]> msgs) {
                             keyAndSchema.schema(), keyAndSchema.value(),
                             valueAndSchema.schema(), valueAndSchema.value(),
-                            msg.offset())
+                            msg.offset(), msg.timestamp(), msg.timestampType())
```
Don't we need to be careful about the timestamp? I think since `ConsumerRecord` returns a `long` for this, couldn't it be the sentinel value `-1` if there is no timestamp? I think we need to check for `NO_TIMESTAMP` or if `timestampType()` is `NO_TIMESTAMP_TYPE`.

shikhar replied:
Yes, it could be the sentinel value if `timestampType() == NO_TIMESTAMP_TYPE`. If the connector makes sure to check `timestampType()` before using the `timestamp()`, it would be fine, but it's probably better to provide a consistent expectation of an unavailable timestamp being `null`, so I can fix this.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java:
```diff
@@ -348,6 +350,30 @@ public void testWakeupInCommitSyncCausesRetry() throws Exception {
+
+        SinkRecord record = records.getValue().iterator().next();
+
+        assertEquals(TIMESTAMP, record.timestamp());
```
Not sure how easy this is with current test code setup, but it seems like we might need to verify translation of a couple of different timestamps, i.e. at least adding the case of no timestamp. I think this is just going to be annoying since `expectConsumerPoll` currently uses a bunch of fixed values.

shikhar replied:
I added a new test for the missing timestamp case (`testMissingTimestampPropagation()`)

ewencp commented:
@shikhar Looks like this is converging, just minor fixes and some edge cases to make sure we have right.

ewencp commented:
LGTM, thanks!
